src/coreclr/nativeaot/System.Private.CoreLib/src/System/Diagnostics/Eventing/EventPipe.NativeAot.cs

```diff
@@ -0,0 +1,162 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+using System.Runtime;
+using System.Runtime.InteropServices;
+
+#if FEATURE_PERFTRACING
+
+namespace System.Diagnostics.Tracing
+{
+    //
+    // NOTE:
+    //
+    // The implementation below takes some manual marshaling actions to ensure arguments are in
+    // primitive form before they are passed through to the underlying RuntimeImports.Rh*
+    // function.
+    //
+    // These extra steps are necessary only if the RuntimeImports mechanism represents "raw"
+    // calls into the native runtime (as has been the case at least in the distant past).
+    //
+    // If the RuntimeImports mechanism automatically applies rich p/invoke marshaling to all of
```

Comment:
Just a note that since these are all LibraryImport, the LibraryImport source generator could provide the necessary marshalling, but I'm not sure it would be low level enough for the purposes of this; I don't have a good sense of when any of this code runs.